text/0031-integ-testing.md

```diff
@@ -0,0 +1,483 @@
+
+```ts
+import { Code, Function, Runtime } from '@aws-cdk/aws-lambda';
+
+// Checks that a standard header is present in the given S3 object
+class StdHeaderAssertion extends Function {
+  constructor(scope: Construct, id: string, props: StdHeaderAssertionProps) {
+    super(scope, id, {
+      code: Code.fromAsset('./standard-header-assertion.js'),
+      runtime: Runtime.NODEJS_12_X,
+      env: {
+        BUCKET_NAME: props.bucket.bucketName,
+        KEY: props.bucket.key
+      },
+    });
+  }
+}
+```
+
+The assertion passes or fails based on the success or failure of the Lambda's handler code.
+In both cases, it can also return an optional message, that will be made available in the test result.
+
+To invoke this assertion, you can simply use the `invokeLambda()` API -
+
+```ts
+const hdrAssertion = new StdHeaderAssertion(...);
+test.invokeLambda(hdrAssertion);
+```
+
+> RFC Note: See [Appendix A](#appendix-a---assertions) for details on the underlying design.
+
+### Test snapshots
+
+On first run of a apptest test, a snapshot will be produced that *must be checked into source tree*.
+This is a snapshot of the [cloud assembly] containing both the CDK app (being tested) and the test
+assertions. The snapshot will be placed at a folder relative to the test file.
+
+```
+test
+├── repo.apptest.ts
+└── repo.apptest.snap
+    ├── manifest.json
+    ├── RepoStack.template.json
+    ├── ...
+    └── ...
+```
+
+On subsequent runs of this test, if the synthesized cloud assembly matches the
+snapshot, apptest will skip deployment and will not run any assertions. The test will be considered
+to have passed. This behaviour can be overridden either by setting the `ignoreSnapshot` option for
+individual tests or by setting the `CDK_APPTEST_IGNORE_SNAPSHOT` environment variable and will apply
+to all tests executed in that environment.
+
+However, when the synthesized cloud assembly does not match the snapshot, the test will be executed
+in full and the snapshot will be updated. As before, the changes to the snapshots must be checked into
+the source tree.
+
+[cloud assembly]: https://docs.aws.amazon.com/cdk/api/latest/docs/cloud-assembly-schema-readme.html#cloud-assembly
+
+### Test Execution & Report
+
+The `apptest` suite can be executed via any standard test runner available in the programming
+language of your choice - jest for javascript or typescript, junit for Java, pytest for Python, etc.
+
+As shown in the above section, it is recommended to use the `apptest` suffix to your test files
+to differentiate them from unit and other kinds of tests - `xxx.apptest.ts` in a typescript
+project, `xxx.apptest.java` for Java project, etc.
+
+This setup will enable defining a separate script that runs the apptest suite - npm script,
+Ant target, Maven phase, etc. - by using `*.apptest.*` as the file filter in your test runner.
+
+The test will be executed when the `run()` method is called on the `apptest.Test` class.
```

Comment:
jsii (currently) does not support throwing error messages across languages.

Reply:
Not sure what you mean by that? You can throw an exception in JavaScript and it will propagate to the host language...